### DISCUSSION TITLE
Where to go next? 🤔

### DISCUSSION THREAD

**Discussion by matthewmueller** [livebud/bud#177] [Proposals]
Now that the codebase has settled in, where do you think we should go next with Bud?

[![](https://api.gh-polls.com/poll/01G6S8M62G346N7H0QEX0XQSM8/Environments%2C%20Models%2C%20Migrations%2C%20Custom%20Routing)](https://api.gh-polls.com/poll/01G6S8M62G346N7H0QEX0XQSM8/Environments%2C%20Models%2C%20Migrations%2C%20Custom%20Routing/vote)

[![](https://api.gh-polls.com/poll/01G6S8M62G346N7H0QEX0XQSM8/Sessions%2C%20Layouts%2C%20Validation%2C%20Type-safe%20Props)](https://api.gh-polls.com/poll/01G6S8M62G346N7H0QEX0XQSM8/Sessions%2C%20Layouts%2C%20Validation%2C%20Type-safe%20Props/vote)

[![](https://api.gh-polls.com/poll/01G6S8M62G346N7H0QEX0XQSM8/Someplace%20Else%3F)](https://api.gh-polls.com/poll/01G6S8M62G346N7H0QEX0XQSM8/Someplace%20Else%3F/vote)

The current plan is to follow the [Roadmap](https://github.com/livebud/bud/discussions/9). I think it's a good plan, but I'm wondering if it makes more sense to double-down on what's already available vs. expanding what's available. 

One reason for expanding is that I've already built the [basic migration system](https://github.com/Matthewmueller/migrate) as well as the generated ORM ([very old product demo](https://www.youtube.com/watch?v=6sQIHcy5nQo)). 

I'm questioning the current roadmap a bit though because there's already a decent [workaround for environments](https://github.com/livebud/bud/discussions/72#discussioncomment-2788198). @vito has also shown us that [XO](https://github.com/xo/xo) already works with https://github.com/vito/bass-loop and that it's already possible to [deploy a Bud app](https://loop.bass-lang.org/).

Why it might be worth doubling down on existing features is that you can't easily add layouts and sessions. Also how props get passed to the views can be very mysterious at times. It'd be nice to make this all type-safe by generating TS or jsdoc type definitions. 

There's no question that both are necessary, it's more about what order unlocks more value for you earlier?

What do you think?

**Comment by alwindoss** [comment]
There should be unified developer experience on all Operating Systems that Go supports.

**Reply by matthewmueller** [reply]
@alwindoss thanks for your comment. I think it works everywhere except Windows. Are you waiting for Windows support?

If so, the main blocker there is V8. See https://github.com/livebud/bud/discussions/81 for more details.

**Reply by alwindoss** [reply]
You are right @matthewmueller . All i am saying is that that problem should be solved. Until that is solved in the upstream Go perhaps a convenience script which is opaque to the user when they install bud.
When it is fixed in the upstream we should we able to remove this extra step without any impact to user.

**Reply by matthewmueller** [reply]
I wish it were that easy. V8 is pretty core to Bud, so there's no way to just emulate V8 until there's support unless we build a JS evaluator! In the meantime WSL2 will have to suffice. What does still need to happen is #130.

**Comment by jfmario** [comment]
In my opinion Sessions are the most pressing feature. When I did some experimenting, I am able to use DI to get a lot done, and was using dbmate (https://github.com/amacneil/dbmate) for migrations. Anything having to do with session was more difficult. I think putting the basics in place that allow for people to build authentication is a good next goal.

**Reply by matthewmueller** [reply]
I'm definitely leaning in this direction. At least with session support. It does seem like there's a desire to go the other way too, so it might make more sense to cycle between improving and expanding.

**Comment by vito** [comment]
My $.02:

Mostly aligned with @jfmario.

Figuring out models/migrations was pretty easy. I just had to turn my existing database setup into a dependency, which already feels intuitive. I'm curious what plans you have anyway. One wrinkle is I'm pretty bullish about SQLite these days and I'd like to keep that. So far all the plans mention PostgreSQL but maybe it's just an example user story focus. :)

My app just needs a tiny bit of local state to track the browser's preferred color scheme. If sessions could help there somehow that would be awesome but I'm not blocked on it or anything, since I don't actually have user sessions (yet?). But considering this is much harder to do "on your own" it would make sense to me to address it sooner than models/etc.

**Reply by matthewmueller** [reply]
> One wrinkle is I'm pretty bullish about SQLite these days and I'd like to keep that. So far all the plans mention PostgreSQL but maybe it's just an example user story focus. :)

Me too! PostgreSQL will be the focus for the beginning though :) 

> But considering this is much harder to do "on your own" it would make sense to me to address it sooner than models/etc.

Yep exactly. I think I've settled that I'll first address the hard/impossible things to do in Bud right now and then get back to adding the M in MVC

---

Question for you @vito: how are you deploying and hosting https://github.com/vito/bass-loop at the moment?

**Reply by vito** [reply]
It's a YOLO'd VM in DigitalOcean at the moment. I just have a script that does `bud build`, sets env vars, and runs `./bud/app`.

**Comment by Zvika88** [comment]
Came across Bud this weekend and am playing with the code and examples. It looks awesome. Kudos @matthewmueller!
What's next? 3 themes, all of which have been mentioned:

- ORM, with focus on Postgres. Document DBs down the road.
- Authentication, authorization, JWT, sessions
- React frontend, and hopefully SSR down the road.
